redash/query_runner/pg.py

```diff
@@ -260,6 +260,10 @@ def run_query(self, query, user):
         cursor = connection.cursor()
 
         try:
+            db_role = getattr(user, "db_role", None)
+            if db_role:
+                logger.info(f"Setting DB role for query: {user.db_role}")
+                cursor.execute("set role %s", user.db_role)
```

Comment:
What if we used a context manager around the query execution? Something like:

```python
@contextmanager
def use_role(cursor, role_name: str):
    if role_name:
        cursor.execute(f"SET ROLE {role_name}")
    yield cursor
    if role_name:
        cursor.execute("RESET ROLE")
```

Reply:
I don't think it's really necessary, since the connection is closed just a bit further down.

Reply:
It seems this is leading to this error:

```
execute cannot be used while an asynchronous query is underway
```

I'm not entirely sure why and need to debug some more, but I may need to change how this works anyway, after all.

Reply:
I got past the async query issue but per discussion in Slack the `set role` approach doesn't really work here because we're allowing arbitrary SQL which could just start with a `RESET ROLE;`. Going to move forward with this for right now but create another ticket to switch the roles to allow for actual login.